### PR TITLE
internal: Add `envelope.{Client, Server}`

### PR DESCRIPTION
**Pull request by abhinav**
`envelope.{Client, Server}` rely on primitive transport interfaces
(implemented by `internal/frame` in #176) to send and receive enveloped Thrift
requests and responses. Out-of-order of requests is not supported.

@breerly @prashantv @kriskowal

### CODE REVIEW

File: internal/envelope/mock_client_test.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
+// Automatically generated by MockGen. DO NOT EDIT!
```
Comment (breerly) [inline internal/envelope/mock_client_test.go]: Github doesn't know how to collapse deez? That's a shame.

File: internal/envelope/client.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+	default:
+		return wire.Value{}, errUnknownEnvelopeType(resEnvelope.Type)
+	}
+}
+
+type errUnknownEnvelopeType wire.EnvelopeType
+
+func (e errUnknownEnvelopeType) Error() string {
+	return fmt.Sprintf("unknown envelope type: expected Reply, got %v", wire.EnvelopeType(e))
```
Comment (breerly) [inline internal/envelope/client.go]: Perhaps "expected `wire.Reply`: ...` here would be more obvious?

**Comment by breerly** [comment]
lgtm (thanks for breaking deez up)

**Comment by prashantv** [comment]
lgtm

**Pull request merged**